src/rate_limiter.rs

```diff
@@ -35,9 +117,42 @@ impl<'a, K: Eq + Hash> RateLimiter<'a, K> {
     }
 }
 
-impl<'a, K: Eq + Hash> Default for RateLimiter<'a, K> {
-    fn default() -> Self {
-        Self::new()
+/// The builder exposes ability to configure a [`RateLimiter`] instance by
+/// setting limiting policies. One does not need to create a builder instance
+/// directly, as it must be created indirectly via [`RateLimiter::new`].
+pub struct RateLimiterBuilder<'a, K> {
+    limits: Vec<(K, usize, Duration)>,
+    clock: &'a (dyn Fn() -> Instant + Sync),
+}
+
+impl<'a, K> RateLimiterBuilder<'a, K> {
+    /// Sets a limiting policy for a `key`.
+    ///
+    /// The limiting policy sets how many times an event is allowed to happen
+    /// (`limit`) within a given period of time (`interval`). Event is vague
+    /// term. Thus we use a `key` to uniquely identify an event we want to rate
+    /// limit.
+    pub fn limit(mut self, key: K, limit: usize, interval: Duration) -> Self {
+        self.limits.push((key, limit, interval));
+        self
+    }
+}
+
+impl<'a, K: Eq + Hash> RateLimiterBuilder<'a, K> {
+    /// Constructs a [`RateLimiter`] instance with configured limiting policies.
+    ///
+    /// Once constructed, the `RateLimiter` instance cannot be changed.
+    pub fn build(self) -> RateLimiter<'a, K> {
+        RateLimiter {
+            buckets: self
+                .limits
+                .into_iter()
+                .fold(HashMap::new(), move |mut buckets, limit| {
```

Comment:
You could also do something like:

```rust
self
   .limits
   .into_iter()
   .map(|(key, limit, interval)| (key, TokenBucket::with:clock(limit, interval)))
   .collect()
```

A minor benefit of doing it this way is that a `HashMap` will be created via `with_capacity()` because `collect()` will know the size of the iterator.